server/protocols/jmap-draft-integration-testing/jmap-draft-integration-testing-common/src/test/java/org/apache/james/jmap/draft/methods/integration/SetMailboxesMethodTest.java

```diff
@@ -366,6 +367,107 @@ public void subscriptionUserShouldBeChangedWhenUpdateMailbox() throws Exception
+        createSubMailBox(thirdMailBoxName, secondMailboxId);
+        String thirdMailboxId = mailboxProbe.getMailboxId(MailboxConstants.USER_NAMESPACE, username.asString(), "root.second.third").serialize();
+
+        String fourthMailboxName = "fourth";
+        createSubMailBox(fourthMailboxName, thirdMailboxId);
+
+        String requestBody =
+            "[" +
+                "  [ \"setMailboxes\"," +
+                "    {" +
+                "      \"update\": {" +
+                "        \"" + thirdMailboxId + "\" : {" +
+                "          \"name\" : \"thirdtest\" ,"  +
+                "          \"parentId\" : \"" + secondMailboxId + "\""  +
+                "        }" +
+                "      }" +
+                "    }," +
+                "    \"#0\"" +
+                "  ]" +
+                "]";
+        with()
+            .header("Authorization", accessToken.asString())
+            .body(requestBody)
+            .post("/jmap");
+
+        assertThat(mailboxProbe.listSubscriptions(username.asString()))
+            .contains("root.second.thirdtest.fourth")
+            .doesNotContain("root.second.third.fourth");
+    }
+
+    @Test
+    public void subscriptionUserShouldBeChangedAllChildWhenUpdateParentMailbox() throws Exception {
+        MailboxId parentId = mailboxProbe.createMailbox(MailboxConstants.USER_NAMESPACE, username.asString(), "root");
+
+        String secondMailboxName = "second";
+        createSubMailBox(secondMailboxName, parentId.serialize());
+        String secondMailboxId = mailboxProbe.getMailboxId(MailboxConstants.USER_NAMESPACE, username.asString(), "root.second").serialize();
+
+        String thirdMailBoxName = "third";
+        createSubMailBox(thirdMailBoxName, secondMailboxId);
+        String thirdMailboxId = mailboxProbe.getMailboxId(MailboxConstants.USER_NAMESPACE, username.asString(), "root.second.third").serialize();
+
+        String fourthMailboxName = "fourth";
+        createSubMailBox(fourthMailboxName, thirdMailboxId);
+
+        String requestBody =
+            "[" +
+                "  [ \"setMailboxes\"," +
+                "    {" +
+                "      \"update\": {" +
+                "        \"" + secondMailboxId + "\" : {" +
+                "          \"name\" : \"secondtest\" ,"  +
+                "          \"parentId\" : \"" + parentId.serialize() + "\""  +
+                "        }" +
+                "      }" +
+                "    }," +
+                "    \"#0\"" +
+                "  ]" +
+                "]";
+        with()
+            .header("Authorization", accessToken.asString())
+            .body(requestBody)
+            .post("/jmap");
+
+        assertThat(mailboxProbe.listSubscriptions(username.asString()))
+            .contains("root.secondtest.third.fourth")
+            .doesNotContain("root.second.third.fourth");
+    }
+
+    private void createSubMailBox(String childMailboxName, String parentMailboxId) {
```

Comment:
I would put parentId first, and I would enforce a strong MailboxId type for it to avoid mistakes...

Reply:
> I would enforce a strong MailboxId type for it to avoid mistakes...

I would prefer simple String for that kind of method only used in specific test, and we passing `existedMailboxId` from other operation result.
It very hard to make mistake here, it too much dont you think?

Reply:
It's still better to strong type it. It doesn't only make the code more robust and avoiding mistakes, it gives a better readability too